### PR TITLE
Backstage: parameterise 'test' jobs and use Github Action's matrix

### PR DESCRIPTION
**Pull request by frshwtr**
## Describe your changes
Parameterise `test` job and use Github Action's `matrix` configuration to provision tests.

This has a number of benefits:
* Reduces the lines-of-code in the generated `./github/workflows/test.yml`. Improving readability (though one caveat is there are a lot of paramters now set up).
* Only creates jobs for tests that need to run - previously test jobs would be created but not executed, polluting the 'Checks' view in a PR.
* Preserves existing concurrency

Please **do not merge** this once approved, I have left a change in a component file to demonstrate this working. This will need removing before merging.
## Checklist before requesting a review
- [ ] I have applied `percy` label on my PR before merging and after review.
- [ ] If it is a new feature, I have added thorough tests.
- [ ] I have updated relevant docs.

### CODE REVIEW

File: templates/test-workflow.yml [hunk 1]
```diff
@@ -26,23 +24,29 @@ jobs:
...
-  test-<%& workspaceFilename %>:
+  test:
+    if: ${{ needs.changes.outputs.components != '[]' && needs.changes.outputs.components != '' }}
     needs: changes
-    if: ${{ needs.changes.outputs.<%& workspaceFilename %> == 'true' }}
     runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        component: ${{ fromJson(needs.changes.outputs.components) }}
+        include:
```
[the remark this answers is not in the document]
Comment (frshwtr) [reply templates/test-workflow.yml:34]: woaahh, that's definitely not right. I'll look into this.

[the remark this answers is not in the document]
Comment (frshwtr) [reply templates/test-workflow.yml]: @notlee fixed, a simpler approach was to change the generated filter group name in `paths_filter` to be the path. This actually simplifies the script further which is an added bonus :)

**Review by notlee — APPROVED** [review]

**Comment by frshwtr** [comment]
I'll add my responses from Slack for transparency.

> One to answer before merging:
> Are you sure it will let us rerun individual failed tests, say if only 1 out of 50 fails? Or should we test that now?

Tested and confirmed. Added `fail-fast` which will prevent other tests in the matrix from being cancelled if one fails.

> And just a thought:
> Think we could apply the same to Percy? Probably as a separate PR

Absolutely! We might need to utilise a custom job that runs our "should I run percy" scripts to generate a list of percy jobs to trigger, and pass that output to Github Actions' output in a format that can be ingested by a matrix.

image

**Comment by notlee** [comment]
As discussed: Final thing to work out before merging is how to keep branch protections.

**Pull request merged**